aten/src/ATen/native/ReduceOps.cpp

```diff
@@ -276,32 +299,49 @@ std::tuple<Tensor&, Tensor&> cummin_out(Tensor& values, Tensor& indices, const T
     values.resize_(self.sizes());
     indices.resize_(self.sizes());
     if(self.dim() == 0) {
-      values.fill_(self.item());
+      values.fill_(self);
       indices.fill_(0);
     }
     else if(self.numel() != 0) {
-      // update values and indices for the first values along the dimension dim
-      values.narrow(dim, 0, 1) = self.narrow(dim, 0, 1);
-      indices.narrow(dim, 0, 1).fill_(0);
-      for(int i = 1; i < self.size(dim); i++) {
-        auto res_at_i = at::min(at::cat({values.narrow(dim, i-1, 1), self.narrow(dim, i, 1)}, dim), dim, true);
-        // values at index i
-        values.narrow(dim, i, 1) = std::get<0>(res_at_i);
-        // indices at index i
-        indices.narrow(dim, i, 1) = at::max(indices.narrow(dim, i-1, 1), (i * (std::get<1>(res_at_i))));
-       }
+      AT_DISPATCH_ALL_TYPES(self.scalar_type(), "cummin", [&]() -> void {
+        auto values__data = values.data_ptr<scalar_t>();
+        auto self__data = self.data_ptr<scalar_t>();
+        int64_t values__stride_dim = values.stride(dim);
+        int64_t self__stride_dim = self.stride(dim);
+        at::parallel_for(0, self.size(dim), internal::GRAIN_SIZE, [&](int64_t p_begin, int64_t p_end) {
+          values__data[0] = self__data[0];
+          auto cummin = values__data[0];
+          auto idx = 0;
+          AT_DISPATCH_ALL_TYPES(kLong, "cummin_idx_0", [&]() -> void {
+            auto indices__data = indices.data_ptr<scalar_t>();
+            indices__data[0] = 0;
+          });
+          for(int64_t i = p_begin + 1; i < p_end; i++) {
+            if(self__data[i*self__stride_dim] <= cummin) {
```

Comment:
I don't think you can parallelize scan operations with a regular paralel_for loop, because the result depends on all previous values, and your current thread has seen only some of them. You start by comparing with cummin which was set to the very first value (at index 0), but what you should be comparing against is the result of the scan operation for all the previous values, which is not easy to obtain in a parallel fashion. I recommend looking at how cumprod and cumsum are currently implemented in TH/THC.

Reply:
yeah I agree with what you are saying. updated the code and added benchmark results in the description